presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java

```diff
@@ -379,26 +379,39 @@ private PlanBuilder aggregateGroupingSet(List<FieldOrExpression> groupingSet, Pl
                 rewritten = ((Cast) rewritten).getExpression();
                 needPostProjectionCoercion = true;
             }
-            aggregationAssignments.put(newSymbol, (FunctionCall) rewritten);
+            aggregationAssignmentsBuilder.put(newSymbol, (FunctionCall) rewritten);
             translations.put(aggregate, newSymbol);
 
-            functions.put(newSymbol, analysis.getFunctionSignature(aggregate));
+            functionsBuilder.put(newSymbol, analysis.getFunctionSignature(aggregate));
         }
 
+        Map<Symbol, FunctionCall> aggregationAssignments = aggregationAssignmentsBuilder.build();
+        Map<Symbol, Signature> functions = functionsBuilder.build();
+
         // 2.b. Rewrite group by expressions in terms of pre-projected inputs
-        Set<Symbol> groupBySymbols = new LinkedHashSet<>();
-        for (FieldOrExpression fieldOrExpression : groupingSet) {
-            Symbol symbol = subPlan.translate(fieldOrExpression);
-            groupBySymbols.add(symbol);
-            translations.put(fieldOrExpression, symbol);
+        ImmutableList.Builder<List<Symbol>> groupingSetsSymbolsBuilder = ImmutableList.builder();
+        ImmutableList.Builder<Symbol> allGroupingSymbolsBuilder = ImmutableList.builder();
+        for (List<FieldOrExpression> groupingSet : groupingSets) {
+            ImmutableList.Builder<Symbol> groupingColumns = ImmutableList.builder();
+            for (FieldOrExpression fieldOrExpression : groupingSet) {
+                Symbol symbol = subPlan.translate(fieldOrExpression);
+                groupingColumns.add(symbol);
+                allGroupingSymbolsBuilder.add(symbol);
+                translations.put(fieldOrExpression, symbol);
+            }
+            groupingSetsSymbolsBuilder.add(groupingColumns.build());
         }
+        List<List<Symbol>> groupingSetsSymbols = groupingSetsSymbolsBuilder.build();
+        List<Symbol> allGroupingSymbols = allGroupingSymbolsBuilder.build();
 
         // 2.c. Mark distinct rows for each aggregate that has DISTINCT
         // Map from aggregate function arguments to marker symbols, so that we can reuse the markers, if two aggregates have the same argument
         Map<Set<Expression>, Symbol> argumentMarkers = new HashMap<>();
         // Map from aggregate functions to marker symbols
         Map<Symbol, Symbol> masks = new HashMap<>();
-        for (FunctionCall aggregate : Iterables.filter(analysis.getAggregates(node), FunctionCall::isDistinct)) {
+        for (FunctionCall aggregate : analysis.getAggregates(node).stream()
+                .filter(FunctionCall::isDistinct)
+                .collect(toImmutableList())) {
```

Comment:
It seems very convoluted to do filter and collect in for-each. I understand you rewrote this because the old code uses pre-8 style functional call. I'd rewrite this to `if ... break`